### PR TITLE
Improve the error message for invalid table key values

### LINKED ISSUE #35043: No error when non-constant expressions are used as values for key fields
**Description:**
$title.

**Steps to reproduce:**
```ballerina
type Employee record {|
    readonly int|string|xml|boolean id;
    readonly string name;
    float salary?;
|};

const CONST = "constant";

function init() {
    int i = 1;
    boolean b = false;

    table<Employee> _ = table key(id) [
        // {id: i, name: ""}, // @error a constant expression is required as the value for key field 'id' - error is logged
        {id: string `foo${i}`, name: "Em"}, // @error a constant expression is required as the value for key field 'id'
        // {id: xml `xml text ${i}`, name: "Em"}, // @error a constant expression is required as the value for key field 'id' - different error is logged
        {id: !b, name: "Em"}, // @error a constant expression is required as the value for key field 'id'
        {id: 1 + i, name: "Amy"} // @error a constant expression is required as the value for key field 'id'
    ];
}
```

**Affected Versions:**
2201.0.0

### PR DESCRIPTION
**Pull request by nipunayf**
## Purpose

The current implementation of checking whether the table key value is a constant expression does not consider its nested expressions. Thus, this PR improves the implementation to precisely capture the `value expression of key specifier 'k' must be a constant expression` error message.

As mentioned in the specification, a constant-reference expression must reference a module-const-decl, and within a const-expr, any nested expression must also be a const-expr. The existing positive table key field value tests do not adhere to this condition, and this PR updates those test cases with the necessary changes. https://ballerina.io/spec/lang/master/#const-expr

Fixes #35043

## Approach
The PR modifies the TypeChecker to improve its implementation of checking the constant expression for table key values.

## Remarks
* The table constructor uses a method named `isConstExpression` that is specific to the table itself. Hence, the changes made by the PR only affect this area. Ideally, this method can be used in other areas as well, and this is tracked with another issue:
* The current implementation does not consider the `list constructor spread operator` as it is blocked by: https://github.com/ballerina-platform/ballerina-lang/issues/41979
* The current implementation does not consider the `xml templates` as it is blocked by: https://github.com/ballerina-platform/ballerina-lang/issues/41985
* The current implementation does not consider the `regex templates` it is blocked by: https://github.com/ballerina-platform/ballerina-lang/issues/41982

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -1522,28 +1522,75 @@ private boolean isConstExpression(BLangExpression expression) {
...
+                        }
+                        case RECORD_LITERAL_SPREAD_OP -> {
+                            if (!isConstExpression(((BLangRecordLiteral.BLangRecordSpreadOperatorField) field).expr)) {
+                                return false;
+                            }
+                        }
+                        case SIMPLE_VARIABLE_REF -> {
+                            if (!isConstExpression(((BLangRecordVarNameField) field))) {
+                                return false;
+                            }
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1573]:
```suggestion
                            return !isConstExpression(((BLangRecordVarNameField) field))
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -1522,28 +1522,75 @@ private boolean isConstExpression(BLangExpression expression) {
...
+                        case RECORD_LITERAL_KEY_VALUE -> {
+                            if (!isConstExpression(((BLangRecordKeyValueField) field).valueExpr) ||
+                                    !isConstExpression(((BLangRecordKeyValueField) field).key.expr)) {
+                                return false;
+                            }
+                        }
+                        case RECORD_LITERAL_SPREAD_OP -> {
+                            if (!isConstExpression(((BLangRecordLiteral.BLangRecordSpreadOperatorField) field).expr)) {
+                                return false;
+                            }
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1568]:
```suggestion
                            return !isConstExpression(((BLangRecordLiteral.BLangRecordSpreadOperatorField) field).expr))
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -1522,28 +1522,75 @@ private boolean isConstExpression(BLangExpression expression) {
...
+            case RECORD_LITERAL_EXPR:
+                BLangRecordLiteral recordLiteral = (BLangRecordLiteral) expression;
+                List<RecordLiteralNode.RecordField> fields = recordLiteral.getFields();
+                for (RecordLiteralNode.RecordField field : fields) {
+                    switch (field.getKind()) {
+                        case RECORD_LITERAL_KEY_VALUE -> {
+                            if (!isConstExpression(((BLangRecordKeyValueField) field).valueExpr) ||
+                                    !isConstExpression(((BLangRecordKeyValueField) field).key.expr)) {
+                                return false;
+                            }
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1563]:
```suggestion
                            return isConstExpression(((BLangRecordKeyValueField) field).valueExpr) ||
                                    isConstExpression(((BLangRecordKeyValueField) field).key.expr)
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 4]
```diff
@@ -1522,28 +1522,75 @@ private boolean isConstExpression(BLangExpression expression) {
...
+                        }
+                        case RECORD_LITERAL_SPREAD_OP -> {
+                            if (!isConstExpression(((BLangRecordLiteral.BLangRecordSpreadOperatorField) field).expr)) {
+                                return false;
```
Comment (KavinduZoysa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1567]: Can you check why isn't this covered?

Comment (nipunayf) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1567]: The current implementation throws an exception, as mentioned here: https://github.com/ballerina-platform/ballerina-lang/issues/41979#issuecomment-1895340353, when using the record spread operator as a constant expression. Hence, we need to resolve this before adding a test case.

**Review by KavinduZoysa — APPROVED** [review]

**Review by KavinduZoysa — APPROVED** [review]

**Pull request merged**